packages/tuono-router/src/components/RouterContext.tsx

```diff
@@ -20,3 +36,88 @@ export function getRouterContext(): React.Context<Router> {
 
   return routerContext
 }
+
+interface RouterContextProviderProps {
+  router: Router
+  children: ReactNode
+  serverSideProps?: ServerRouterInfo
+}
+
+function getInitialLocation(
+  serverSideProps?: ServerRouterInfo,
+): ParsedLocation {
+  if (typeof document === 'undefined') {
+    return {
+      pathname: serverSideProps?.pathname || '',
+      hash: '',
+      href: serverSideProps?.href || '',
+      searchStr: serverSideProps?.searchStr || '',
+      // TODO: Polyfill URLSearchParams
+      search: {},
+    }
+  }
+
+  const { location } = window
+  return {
+    pathname: location.pathname,
+    hash: location.hash,
+    href: location.href,
+    searchStr: location.search,
+    search: Object.fromEntries(new URLSearchParams(location.search)),
+  }
+}
+
+export function RouterContextProvider({
+  router,
+  children,
+  serverSideProps,
+}: RouterContextProviderProps): ReactNode {
+  // Allow the router to update options on the router instance
+  router.update({ ...router.options } as Parameters<typeof router.update>[0])
+
+  const [location, updateLocation] = useState<ParsedLocation>(
+    getInitialLocation(serverSideProps),
+  )
+
+  /**
+   * Listen browser navigation events
+   */
+  useEffect(() => {
+    const updateLocationOnPopStateChange = ({
+      target,
+    }: PopStateEvent): void => {
+      const { location: targetLocation } = target as typeof window
+      const { pathname, hash, href, search } = targetLocation
+
+      updateLocation({
+        pathname,
+        hash,
+        href,
+        searchStr: search,
+        search: Object.fromEntries(new URLSearchParams(search)),
+      })
+    }
+
+    window.addEventListener('popstate', updateLocationOnPopStateChange)
+
+    return (): void => {
+      window.removeEventListener('popstate', updateLocationOnPopStateChange)
+    }
+  }, [updateLocation])
+
+  const RouterContext = getRouterContext()
+
+  const context: RouterContextType = {
+    router,
+    location,
+    updateLocation,
+  }
+
+  return (
+    <RouterContext.Provider value={context}>{children}</RouterContext.Provider>
+  )
+}
+
+export function useInternalRouter(): RouterContextType {
```

Comment:
All names in this file speak of `RouterContext`. The word Internal comes out only here.
Since his hook is not exposed to the user I think we should named it simply `useRouterContext`.
To avoid accidental exports we might add a `@warning` comment

```suggestion
/** @warning DO NOT EXPORT THIS TO USER LAND */
export function useRouterContext(): RouterContextValue {
```